Tribler/community/multichain/community.py

```diff
@@ -0,0 +1,251 @@
+                     "a81fdcb2db302ac15905d6e75f96b39ccdaf068bdbbda81a6356f53f7ce4e"
+        master_key_hex = master_key.decode("HEX")
+        master = dispersy.get_member(public_key=master_key_hex)
+        return [master]
+
+    def initiate_meta_messages(self):
+        return super(MultiChainCommunity, self).initiate_meta_messages() + [
+            Message(self, SIGNATURE,
+                    DoubleMemberAuthentication(
+                        allow_signature_func=self.allow_signature_request, split_payload_func=split_function),
+                    PublicResolution(),
+                    DirectDistribution(),
+                    CandidateDestination(),
+                    SignaturePayload(),
+                    self._generic_timeline_check,
+                    self.received_signature_response)]
+
+    def initiate_conversions(self):
+        return [DefaultConversion(self), MultiChainConversion(self)]
+
+    def publish_signature_request_message(self, candidate):
+        """
+        Creates and sends out a signed signature_request message.
+        """
+        self._logger.info("Sending signature request.")
+        # The lock is lifted after the timeout or a valid signature response is received.
+        self._logger.info("Get Lock: send signature request: %s" % self.chain_lock.locked())
+        self.chain_lock.acquire()
+        self._logger.info("Acquired Lock: send signature request.")
+
+        message = self.create_signature_request_message(candidate)
+        self.create_signature_request(candidate, message, self.allow_signature_response, timeout=5.0)
+
+    def create_signature_request_message(self, candidate):
+        """
+        Create a signature request message using the current time stamp.
+        :return: Signature_request message ready for distribution.
+        """
+        # Instantiate the data
+        up = 1
+        down = 2
+        total_up = 3
+        total_down = 4
+        # Instantiate the personal information
+        sequence_number_requester = self._get_next_sequence_number()
+        previous_hash_requester = self._get_latest_hash()
+
+        payload = (up, down, total_up, total_down, sequence_number_requester, previous_hash_requester)
+        meta = self.get_meta_message(SIGNATURE)
+
+        message = meta.impl(authentication=([self.my_member, candidate.get_member()],),
+                            distribution=(self.claim_global_time(),),
+                            payload=payload)
+        return message
+
+    def allow_signature_request(self, message):
+        """
+        We've received a signature request message, we must either:
+            a. append to this message our data (Afterwards we sign the message.).
+            b. None (if we want to drop this message)
+        """
+        self._logger.info("Received signature request.")
+        self._logger.info("Chain Lock: process request: %s" % self.chain_lock.locked())
+        if self.chain_lock.acquire(False):
```

Comment:
There is no data loss in the tests, as all peers run at the same machine.

So basically, what you're attempting to prevent is A creating two chains with different peers at the same time. Right?

Why would you need a lock for that? My guess is that you're going to call publish_signature_request_message periodically right? If so, set the interval higher than the timeout and you're done.

Moreover, a lock won't prevent branching. If for some reason the message singed by Bob is lost, Alice will have a timeout, create another message with the same hash and hence a branch.

My guess is this can only be prevented by letting Alice sign the message last, preventing a branch if the message from Bob is lost. E.g. use the orginal doublesignedmember auth thingy